psl/psl-core/src/validate/validation_pipeline/validations/relation_fields.rs

```diff
@@ -140,39 +142,79 @@ pub(super) fn ignored_related_model(field: RelationFieldWalker<'_>, ctx: &mut Co
 pub(super) fn referential_actions(field: RelationFieldWalker<'_>, ctx: &mut Context<'_>) {
     let connector = ctx.connector;
     let relation_mode = ctx.relation_mode;
-    let msg = |action: ReferentialAction| {
-        let allowed_values = connector
-            .referential_actions(&relation_mode)
-            .iter()
-            .map(|f| format!("`{}`", f.as_str()))
-            .join(", ");
+
+    fn fmt_allowed_actions(allowed_actions: BitFlags<ReferentialAction>) -> String {
+        allowed_actions.iter().map(|f| format!("`{}`", f.as_str())).join(", ")
+    }
+
+    // validation template for relationMode = "foreignKeys"
+    let msg_foreign_keys = |action: ReferentialAction| {
+        let allowed_actions = connector.referential_actions(&relation_mode);
 
         format!(
             "Invalid referential action: `{}`. Allowed values: ({})",
             action.as_str(),
-            allowed_values,
+            fmt_allowed_actions(allowed_actions),
+        )
+    };
+
+    // validation template for relationMode = "prisma"
+    let msg_prisma = |action: ReferentialAction| {
+        let allowed_actions = connector.emulated_referential_actions(&relation_mode);
+
+        let additional_info = match action {
+            ReferentialAction::NoAction => {
+                if ctx
+                    .connector
+                    .supports_referential_action(&relation_mode, ReferentialAction::Restrict)
+                {
+                    Some(format!(
+                        ". `{}` is not implemented for {} when using `relationMode = \"prisma\"`, you could try using `{}`, which behaves the same if you do not need to defer constraint checks in a transaction. Learn more at https://pris.ly/d/relationMode",
```

Comment:
I'm a bit wary of "which behaves the same if you do not need to defer constraint checks in a transaction" — I don't think we are in a state where we can promise that in the query engine at the moment.

Reply:
With https://github.com/prisma/prisma-engines/pull/3276, we should reach that point, as `NoAction` will effectively be disabled for `relationMode = "prisma"`

Reply:
(We hope both of these PRs can be merged by tomorrow)

Reply:
I meant the query engine doesn't replicate database behaviour, not that its behaviour on restrict and noaction are the same.

If we remove NoAction when the relation mode is "prisma", in that linked PR, does this error still make sense?

Reply:
> I meant the query engine doesn't replicate database behaviour.

Correct, with `relationMode = "prisma"` it emulates its referential actions. Just like `NO ACTION` is pretty close to `RESTRICT` in most SQL databases when using foreign keys, in our emulation we define `NoAction` as an alias of `Restrict` (https://github.com/prisma/prisma-engines/pull/3276, before this PR `NoAction` wouldn't do anything at all, which diverges greatly from what the databases usually do in practice).

> If we remove NoAction when the relation mode is "prisma", in that linked PR, does this error still make sense?

`NoAction` is not removed, it's transparently aliased to `Restrict` in the linked PR.
For Postgres and sqlite, which used to support a buggy emulated `NoAction` referential action, we now suggest them to use `Restrict` with `relationMode = "prisma"`.

Reply:
The deferring of check constraints in transactions doesn't apply to schemas where relationMode = "prisma", so it's still not clear to me what "the same behaviour" means. To put my mental picture in a diagram:

```
Restrict in the database ≠ Restrict in QE
       ≠						=
NoAction in the database ≠ NoAction in QE

```

But since `NoAction` is forbidden on these connectors when relationMode = prisma, it's not accurate that it behaves the same as Restrict. There is no transparent aliasing on the connectors where that error shows up, if I understand correctly.